### PR TITLE
fixes #24881 - cloud-init nocloud support

### PR DESCRIPTION
**Pull request by timogoebel**
This ports [the foreman_userdata plug-in](https://github.com/theforeman/foreman_userdata) to core.

## Abstract

vSphere offers the functionality to clone vm from templates and customize the templates during cloning. This is very handy for changing the network settings of a newly cloned VM. Foreman fully supports this feature via a userdata template that is converted to cloud-init. The amount of things vSphere allows a user to customize are very limited, though. It's not possible to run a finish script or setup puppet with valid certificates on the new VM. This can be worked around by setting up a two step process: vSphere customization is just used to set up the network config of the host. Cloud-init is used to do the rest of the customization on first boot.

As part of this PR, Foreman should be able to serve cloud-init templates.

This leads to this workflow:
1. An administrator creates a vSphere VM template and makes sure the cloud-init application is installed in the template VM as described below in the client setup section.
2. Foreman creates a new cloned VM in vSphere and passes the userdata template (UserData open-vm-tools) to vSphere if the template is properly assigned to the host.
3. vSphere changes the network settings of the new host as defined in the userdata template and boots the new VM.
4. The VM runs cloud-init and asks Foreman for the cloud-init template (CloudInit default) if the template has been properly assinged to the host. Your VM needs network access to Foreman on ports `tcp/80` and `tcp/443`. Cloud-init then runs the template and executes the defined actions.
5. Cloud-init signals Foreman that the host has been built successfully.

## Client setup

On RHEL7 using cloud-init from EPEL:

```
yum install cloud-init -y

cat << EOF > /etc/cloud/cloud.cfg.d/10_foreman.cfg
datasource_list: [NoCloud]
datasource:
  NoCloud:
    seedfrom: http://foreman.example.com/userdata/
EOF
```

## Client debug

```
# Purge all cloud-init data
rm -rf /var/lib/cloud/*

# Run in foreground with debug mode enabled
/usr/bin/cloud-init -d init
```

## Development

To test this plugin manually during development, you can request a template for a specific host by spoofing the host's IP address via a request header.

```
curl -D - -H 'X-FORWARDED-FOR: 192.168.1.1' http://localhost:3000/userdata/user-data
```

## TODOs

* [x] Wait for #6009 and dry up the code with unattended controller
* [x] Make sure this works with smart-proxy's templates module (https://github.com/theforeman/smart-proxy/pull/606)
* [x] open a PR at community-templates to add the templates there (https://github.com/theforeman/community-templates/pull/514)
* [x] Add a helper to the subnet module that returns a list of DNS servers (#6541)
* [ ] Document this behavior in the Foreman manual (after merge)

cc: @lzap, @sean797

### CODE REVIEW

**Comment by lzap** [comment]
> 2\. UserData open-vm-tools

Not sure what you mean by this, I mean I understand what open-vm-tools is but failing to understand context of this step. Just trying to fully understand the workflow for my testing, I barely know VMWare.

> specific host by spoofing

Call the police! :-)

Allright, I am going to do initial review of the code. Thanks for the feature!

File: app/controllers/userdata_controller.rb
Comment (lzap) [inline app/controllers/userdata_controller.rb:1]: So the plan is to wait until we merge unattended controller changes and then to move this logic into it?

File: app/models/template_kind.rb
Comment (lzap) [inline app/models/template_kind.rb]: One UX thing - now we end up with user_data and cloud-init. We need to somehow explain to users the difference between the two. Maybe different naming would help as well? User data was always pretty tough to understand, maybe "seed data". Well maybe not.

Do we need to separate templates for the same format? I mean, eventually you will end up with two same templates in cloud-init format but one as user_data and the other as cloud-init. Is it possible to merge the two so the same template is seeded into VMWare and then acquired via cloud-init client?

File: app/views/unattended/provisioning_templates/user_data/userdata_open_vm_tools.erb
Comment (lzap) [inline app/views/unattended/provisioning_templates/user_data/userdata_open_vm_tools.erb:36]: It looks like this is doable, all root elements from this template are not clashing with the rest in the other one.

On the other hand, I assumed that open-vm client will silently ignore elements it does not want. This is typical way of parsing YAML, but that might not be the case. Then we still need two separate templates.

File: db/seeds.d/020-provisioning_templates_list.rb
Comment (lzap) [inline db/seeds.d/020-provisioning_templates_list.rb]: We definitely want to have snapshots for these once merged in core :-)

**Review by lzap — COMMENTED** [review]
Looks good.

File: app/controllers/userdata_controller.rb
Comment (timogoebel) [reply app/controllers/userdata_controller.rb:1]: I don't think we should actually merge the controllers, but this controller shares a lot of similar code with unattended controller. We have extracted most of the duplicate code out of the controller, so we should be able to reuse that fairly easily.
I know, you don't like the separate controller, but I think unattended controller is already way to complex. I prefer a nice and small controller for the user data stuff.

File: app/models/template_kind.rb
Comment (timogoebel) [reply app/models/template_kind.rb]: Yeah, I was thinking the same. I think we should use userdata for the cloud-init templates and find a better way for the `open-vm-tools` customization. Let me think of something...

File: app/views/unattended/provisioning_templates/user_data/userdata_open_vm_tools.erb
Comment (timogoebel) [reply app/views/unattended/provisioning_templates/user_data/userdata_open_vm_tools.erb:36]: Merging the templates might cause even more confusion. I'd prefer to avoid this.

File: db/seeds.d/020-provisioning_templates_list.rb
Comment (timogoebel) [reply db/seeds.d/020-provisioning_templates_list.rb]: The PR (#5986) is ready, we just need someone to hit the button :-)

**Comment by timogoebel** [comment]
> > 2. UserData open-vm-tools
> 
> Not sure what you mean by this, I mean I understand what open-vm-tools is but failing to understand context of this step. Just trying to fully understand the workflow for my testing, I barely know VMWare.

The basic steps are:

* You start with a template (a special kind of vm that acts as the source for cloning) in VMWare.
* Foreman tells vmware to clone a new vm from this template. The [clone request](https://github.com/theforeman/foreman/blob/2b0401fc35c5aa0d4e059aba434ade4aafeae2a9/app/models/compute_resources/foreman/model/vmware.rb#L523) contains the rendered `open-vm-tools userdata` template.
* vSphere clones the VM and applies the [customizationspec](https://pubs.vmware.com/vi3/sdk/ReferenceGuide/vim.vm.customization.Specification.html) to the template.
* the new vm boots. it looks like the original template, just with the requested changes applied.

File: app/controllers/userdata_controller.rb
Comment (lzap) [reply app/controllers/userdata_controller.rb:1]: Can you give me more arguments? Because the way I see this: user-data is just a template, very similar situation like with iPXE template you recently added. Why to bother with two controllers?

File: app/models/template_kind.rb
Comment (lzap) [reply app/models/template_kind.rb]: In the worst case we need to add some helper text to users explaining what is what.

File: app/views/unattended/provisioning_templates/user_data/userdata_open_vm_tools.erb
Comment (lzap) [reply app/views/unattended/provisioning_templates/user_data/userdata_open_vm_tools.erb:36]: That's fair point.

File: app/controllers/userdata_controller.rb
Comment (timogoebel) [reply app/controllers/userdata_controller.rb:1]: Well, the metadata endpoint is not a template and a separate controller does not hurt.

Comment (timogoebel) [reply app/controllers/userdata_controller.rb:1]: Another thing is that the cloud-init templates need to be accessible even when the host is not in build mode iirc. That will even be another exception that we have to patch into the controller.

Comment (lzap) [reply app/controllers/userdata_controller.rb:1]: Well, my argument is security. Unattended controller is special one, it skips all the authorization, authentication, CSRF and other protections. It is essentially open window over HTTP into Foreman. Now, I would love to see just one such a window instead of two in the project. Technically this is a template rendering, except the context is a bit customized.

I don't see this as a blocker tho, if you feel like new controller is the right way then please remove duplicated code as much as possible. Two major ones I see are skipping various filters and actions and the host find mechanism (this has been already refactored in the other PR).

**Comment by timogoebel** [comment]
Sorry for the delay here, I want to change the implementation quite a bit and haven't found the time yet.

File: app/views/unattended/provisioning_templates/user_data/userdata_open_vm_tools.erb
Comment (tchellomello) [inline app/views/unattended/provisioning_templates/user_data/userdata_open_vm_tools.erb]: I suggest using the line below as the DNS primary and secondary is not mandatory and on its absence, it will create an invalid YAML file with the output `[ , ]`. 
```erb
dnsServerList: [<%=  [@host.subnet.dns_primary, @host.subnet.dns_secondary].reject { |c| c.empty? }.join(',') %>]
```

File: app/views/unattended/provisioning_templates/user_data/userdata_open_vm_tools.erb
Comment (tchellomello) [inline app/views/unattended/provisioning_templates/user_data/userdata_open_vm_tools.erb]: Same for this line: 
```erb
dnsServerList: [<%=  [@host.subnet.dns_primary, @host.subnet.dns_secondary].reject { |c| c.empty? }.join(',') %>]
```

File: app/views/unattended/provisioning_templates/user_data/userdata_open_vm_tools.erb
Comment (ekohl) [reply app/views/unattended/provisioning_templates/user_data/userdata_open_vm_tools.erb]: In other places we also use the DNS as a list and reject empty values. I think a helper in Foreman that provides a list of non-empty values (or an empty list) would make sense.

File: app/models/template_kind.rb
Comment (ares) [reply app/models/template_kind.rb]: Would it be too bad to call it open-vm-tools template? And keep userdata for cloudinit only?

File: app/views/unattended/provisioning_templates/cloud_init/cloud_init_default.erb
Comment (ares) [inline app/views/unattended/provisioning_templates/cloud_init/cloud_init_default.erb:1]: This needs to go through community templates too. I guess it helps with the review, but we should sync it from there to make sure, we have the latest version and next sync won't causd regressions.

**Review by ares — COMMENTED** [review]
This is great. If you don't mind, I'd suggest adding one thing for TODO section. Write a blog post or PR to our manual, with similar content you used here.

File: app/views/unattended/provisioning_templates/cloud_init/cloud_init_default.erb
Comment (timogoebel) [reply app/views/unattended/provisioning_templates/cloud_init/cloud_init_default.erb:1]: Isn't this part of https://github.com/theforeman/community-templates/pull/514?

**Comment by timogoebel** [comment]
Rebased, added templates to snapshot tests, dryed up the code with unattended controller.

File: app/views/unattended/provisioning_templates/user_data/userdata_open_vm_tools.erb
Comment (timogoebel) [reply app/views/unattended/provisioning_templates/user_data/userdata_open_vm_tools.erb]: I created https://projects.theforeman.org/issues/26250, PR will follow tonight.

Comment (timogoebel) [reply app/views/unattended/provisioning_templates/user_data/userdata_open_vm_tools.erb]: https://github.com/theforeman/foreman/pull/6541

File: app/controllers/concerns/foreman/controller/ip_from_request_env.rb
Comment (ekohl) [inline app/controllers/concerns/foreman/controller/ip_from_request_env.rb:3]: I've been thinking about using https://api.rubyonrails.org/classes/ActionDispatch/RemoteIp.html so we can rely on the standard request IP mechanism but never really knew how to get started. Probably out of scope for this refactor but I'd like your thoughts on this. It is relevant when we do pick up https://github.com/theforeman/puppet-foreman/pull/677 again.

Also note that then setups might have two IPs in `HTTP_X_FORWARDED_FOR` which the RemoteIp middleware can correctly parse.

Comment (timogoebel) [reply app/controllers/concerns/foreman/controller/ip_from_request_env.rb:3]: I think we should use rails standards if we can, but I concur: This is out of the scope of this PR.

File: app/models/template_kind.rb
Comment (timogoebel) [reply app/models/template_kind.rb]: If we call this `open-vm-tools` template, we'd have to change how the vmware compute ressource loads the customization template. It cannot use the userdata template anymore.
I'm not against this approach, but it would further delay this PR as someome (probably me) has to rewrite the code. I'd do that in a separate PR, though. We'd then change this PR to only render the userdata template in the unattended controller and drop the cloud-init template kind for good.

@dgoetz also might have some ideas here.

File: app/views/unattended/provisioning_templates/cloud_init/cloud_init_default.erb
Comment (ares) [reply app/views/unattended/provisioning_templates/cloud_init/cloud_init_default.erb:1]: Yes, but what i mean is that if that PR changes, we may forgot to update it in this PR too. So I think it would be better to drop it from here and after this is merged, do the sync using `script/sync_templates.sh`. Or do it as part of this PR but just before merge.

File: app/models/template_kind.rb
Comment (dgoetz) [reply app/models/template_kind.rb]: I have a customer at the moment starting with Katello (indeed with Linux as central managed os at all) who is going to use the plugin Timo created for this before now trying to get this into core. His planned (and at the moment only possible) workflow is using VMware Templates, set a static IP with _user_data_ and then do all the other configuration via _cloud_init_, which is not possible with core only. From this experience a big +1 for this going into core.
Another option instead of _cloud_init_ would be _finish_ but _finish_ templates get disabled when enabling _user_data_ for the VMware Template. _finish_ would be more easy to understand for most users especially those new to automated provisioning as it is simply shell scripting. cloud_init requires typically to learn something new, in addition to all other Foreman/Katello things.
So having two very similar templates I also fear confuses people, so at least naming and documentation should be in a way to help avoid confusion. Changing how VMware compute resource is working (or simply renaming the template) I also think will confuse many because it is perhaps the one I see most and so I fear many guides which will not be updated are out there.
So _user_data_ would be send to the Cloud/Virtualization solution and _cloud_init_ would be for querying by on the system installed cloud_init. Would this be the case for all compute resources? If yes, naming it specific like _open-vm-tools_ seems wrong to me. Any naming suggestions based on the communication direction? Or would it be enough to have this as help text, in the header comment in the template and documentation in the manual? I have no real solution but at least we should try to make it as simple for users as possible, many are already confused by all the options and templates Foreman provides.

Comment (timogoebel) [reply app/models/template_kind.rb]: > So user_data would be send to the Cloud/Virtualization solution and cloud_init would be for querying by on the system installed cloud_init.

Yes.

> If yes, naming it specific like open-vm-tools seems wrong to me

I concur.

> would it be enough to have this as help text, in the header comment in the template

I like that idea.

**Comment by timogoebel** [comment]
Rebased. Changed templates to use `dns_server` helper.

File: app/views/unattended/provisioning_templates/cloud_init/cloud_init_default.erb
Comment (timogoebel) [reply app/views/unattended/provisioning_templates/cloud_init/cloud_init_default.erb:1]: I think we should just merge the community-templates PR in sync with this PR. I like having the templates here so we automatically run snapshot & seed tests with the templates.

Comment (lzap) [reply app/views/unattended/provisioning_templates/cloud_init/cloud_init_default.erb:1]: Let's simply agree that templates in this PR are just copies of the c-t repository, that's the master. Just avoid reviewing the templates here in this PR.

File: app/controllers/unattended_controller.rb
Comment (lzap) [inline app/controllers/unattended_controller.rb:83]: This method is defined in the concern and redefined here, is this intentional?

File: app/models/template_kind.rb
Comment (lzap) [reply app/models/template_kind.rb]: I agree that we should better explain template kinds to users and we already do! This was added in 1.20 I think (extra help line / bubble text). Here it is:

```ruby
    @@template_description ||= {
      "PXELinux" => N_("Used when PXELinux loader is set, loads pxelinux.0 which loads content generated by this template."),
      "PXEGrub" => N_("Used when Grub2 loader is set, loads grub/bootx64.efi which loads content generated by this template."),
      "PXEGrub2" => N_("Used when PXELinux loader is set, loads grub/grubx64.efi which loads content generated by this template."),
      "iPXE" => N_("Used in iPXE environments."),
      "provision" => N_("Template for OS installer, for example kickstart, preseed or jumpstart. Depends on the operating system."),
      "finish" => N_("Post-install script for preseed-based or cloud instance. Connection is made via SSH, credentials or key must exist and inventory IP address must match. Only used when 'user data' is not set."),
      "script" => N_("An arbitrary script, must be manually downloaded using wget/curl."),
      "user_data" => N_("Template with seed data for virtual or cloud instances when 'user data' flag is set, typically cloud-init or ignition format."),
      "ZTP" => N_("Provisioning Junos devices (Junos 12.2+)."),
      "POAP" => N_("Provisioning for switches running NX-OS.")
    }
```

I guess we are doing better job now, it's obvious that user_data is explained as "seed data". Finish script is also well covered. I think we are fine here.

File: app/views/unattended/provisioning_templates/cloud_init/cloud_init_default.erb
Comment (lzap) [reply app/views/unattended/provisioning_templates/cloud_init/cloud_init_default.erb:1]: The one who merges this PR must make sure that the c-r repo PR is correct (the same content or it works).

**Review by lzap — DISMISSED** [review]
This is good, I will start testing this the moment @timogoebel removes the WIP tag from the title :-) Thanks a lot of work here.

**Comment by lzap** [comment]
Please let me know when ready, GH will not notify me.

File: app/controllers/unattended_controller.rb
Comment (timogoebel) [reply app/controllers/unattended_controller.rb:83]: Yeah, we need different behavior.

**Comment by ShimShtein** [comment]
@timogoebel, While I have no problem with this PR, I wonder why did you want to put this capability into core. As opposed to leaving it out and making it an opt-out in the installer.

**Comment by timogoebel** [comment]
> @timogoebel, While I have no problem with this PR, I wonder why did you want to put this capability into core. As opposed to leaving it out and making it an opt-out in the installer.

No specific reason. Folks asked me to move this to core. I can perfectly live with this in a plugin.

btw: I'll probably finish this tomorrow night.

**Comment by timogoebel** [comment]
Rebased and re-synced the community-templates PR. I believe this is ready to merge.

**Review by lzap — DISMISSED** [review]
The code looks good, all relevant PRs are ready except smart-proxy templates which needs rebase. We can merge it later on. I am currently testing the feature.

**Comment by lzap** [comment]
A nit pick, template kind should probably start with capital letter in the description and we should provide some text:

image

```diff
diff --git a/app/models/template_kind.rb b/app/models/template_kind.rb
index b4a504645..9cbd198be 100644
--- a/app/models/template_kind.rb
+++ b/app/models/template_kind.rb
@@ -21,7 +21,7 @@ class TemplateKind < ApplicationRecord
       "user_data" => N_("User data template"),
       "ZTP" => N_("ZTP PXE template"),
       "POAP" => N_("POAP PXE template"),
-      "cloud-init" => N_("cloud-init template")
+      "cloud-init" => N_("Cloud-init template")
     }
   end
 
@@ -36,7 +36,8 @@ class TemplateKind < ApplicationRecord
       "script" => N_("An arbitrary script, must be manually downloaded using wget/curl."),
       "user_data" => N_("Template with seed data for virtual or cloud instances when 'user data' flag is set, typically cloud-init or ignition format."),
       "ZTP" => N_("Provisioning Junos devices (Junos 12.2+)."),
-      "POAP" => N_("Provisioning for switches running NX-OS.")
+      "POAP" => N_("Provisioning for switches running NX-OS."),
+      "cloud-init" => N_("Template for cloud-init unattended endpoint.")
     }
   end
```

**Comment by lzap** [comment]
Also the name of the template reads a bit off for me, "CloudInit default". I'd probably go with "Cloud-init default", looks much nicer. Just rename the file as well.

**Comment by lzap** [comment]
I am getting an error when you provide an incorrect IP (non-existing host):

```
[app|F|74c|74c] NameError (undefined local variable or method `error_message' for #<UserdataController:0x0000555561419590>
 | Did you mean?  errors_hash):
[app|F|74c|74c]   
[app|F|74c|74c] app/controllers/concerns/foreman/controller/template_rendering.rb:28:in `render_error'
 | app/controllers/userdata_controller.rb:62:in `find_host'
 | app/controllers/concerns/foreman/controller/timezone.rb:10:in `set_timezone'
 | app/models/concerns/foreman/thread_session.rb:32:in `clear_thread'
 | app/controllers/concerns/foreman/controller/topbar_sweeper.rb:12:in `set_topbar_sweeper_controller'
 | lib/foreman/middleware/telemetry.rb:10:in `call'
 | lib/foreman/middleware/catch_json_parse_errors.rb:9:in `call'
 | lib/foreman/middleware/logging_context_session.rb:22:in `call'
 | lib/foreman/middleware/logging_context_request.rb:11:in `call'
[app|I|74c|74c]   Rendered /home/lzap/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/actionpack-5.2.1/lib/action_dispatch/middleware/templates/rescues/diagnostics.html.erb within rescues/layout (32.6ms)
```

**Comment by lzap** [comment]
Other than that it works fine.

**Comment by timogoebel** [comment]
> I am getting an error when you provide an incorrect IP (non-existing host):

Ah, this needs a test as well. Will take a look.

**Comment by timogoebel** [comment]
This fixes the issue:

```diff
diff --git a/app/controllers/concerns/foreman/controller/template_rendering.rb b/app/controllers/concerns/foreman/controller/template_rendering.rb
index 4f9e0ba23..16662e24b 100644
--- a/app/controllers/concerns/foreman/controller/template_rendering.rb
+++ b/app/controllers/concerns/foreman/controller/template_rendering.rb
@@ -25,9 +25,9 @@ module Foreman
       end

       def render_error(options)
-        logger.error error_message % params
         message = options.delete(:message)
         status = options.delete(:status) || :not_found
+        logger.error message % options
         render plain: "#{message % options}\n", :status => status
       end
     end
diff --git a/app/models/template_kind.rb b/app/models/template_kind.rb
index b4a504645..9cbd198be 100644
--- a/app/models/template_kind.rb
+++ b/app/models/template_kind.rb
@@ -21,7 +21,7 @@ class TemplateKind < ApplicationRecord
       "user_data" => N_("User data template"),
       "ZTP" => N_("ZTP PXE template"),
       "POAP" => N_("POAP PXE template"),
-      "cloud-init" => N_("cloud-init template")
+      "cloud-init" => N_("Cloud-init template")
     }
   end

@@ -36,7 +36,8 @@ class TemplateKind < ApplicationRecord
       "script" => N_("An arbitrary script, must be manually downloaded using wget/curl."),
       "user_data" => N_("Template with seed data for virtual or cloud instances when 'user data' flag is set, typically cloud-init or ignition format."),
       "ZTP" => N_("Provisioning Junos devices (Junos 12.2+)."),
-      "POAP" => N_("Provisioning for switches running NX-OS.")
+      "POAP" => N_("Provisioning for switches running NX-OS."),
+      "cloud-init" => N_("Template for cloud-init unattended endpoint.")
     }
   end

diff --git a/test/controllers/userdata_controller_test.rb b/test/controllers/userdata_controller_test.rb
index 1efb82450..57f619c68 100644
--- a/test/controllers/userdata_controller_test.rb
+++ b/test/controllers/userdata_controller_test.rb
@@ -63,6 +63,15 @@ class UserdataControllerTest < ActionController::TestCase
         assert_response :success
         assert_equal user_data_content, @response.body
       end
+
+      context 'with unknown ip address' do
+        test 'should display an error' do
+          @request.env['REMOTE_ADDR'] = '198.51.100.1'
+          get :userdata
+          assert_response :not_found
+          assert_includes @response.body, 'Could not find host for request 198.51.100.1'
+        end
+      end
     end

     context 'with cloud-init template' do
```

**Comment by timogoebel** [comment]
@lzap: Tests passed.

**Comment by lzap** [comment]
Please rebase.

**Comment by timogoebel** [comment]
Rebased.

**Review by lzap — APPROVED** [review]
Works fine, thank you so much for doing the extra work and pushing this into core. Katello tests did not run however this provisioning endpoint is very unlikely to affect Katello in any way.

**Comment by lzap** [comment]
Merged, thanks. Feel free to turn the description into manual and send it to manual site and/or tutorial section of our discourse.

**Comment by ares** [comment]
Awesome work! I'm happy to see this in core. Also I'm looking forward for the documentation update :-)

**Comment by timogoebel** [comment]
The only thing still missing is the possibility to create a template from an existing host. I believe we already have all the bits and pieces, we just have to glue them together.

**Pull request merged**
